module/zfs/vdev_mirror.c

```diff
@@ -252,16 +252,21 @@ vdev_mirror_map_init(zio_t *zio)
 		dva_t *dva = zio->io_bp->blk_dva;
 		spa_t *spa = zio->io_spa;
 
-		mm = vdev_mirror_map_alloc(BP_GET_NDVAS(zio->io_bp), B_FALSE,
-		    B_TRUE);
+		c = BP_GET_NDVAS(zio->io_bp);
+		if (zio->io_mirror_child_hint != -1)
+			c = zio->io_mirror_child_hint % c + 1;
+		mm = vdev_mirror_map_alloc(c, B_FALSE, B_TRUE);
 		for (c = 0; c < mm->mm_children; c++) {
 			mc = &mm->mm_child[c];
 
 			mc->mc_vd = vdev_lookup_top(spa, DVA_GET_VDEV(&dva[c]));
 			mc->mc_offset = DVA_GET_OFFSET(&dva[c]);
 		}
 	} else {
-		mm = vdev_mirror_map_alloc(vd->vdev_children,
+		c = vd->vdev_children;
+		if (zio->io_mirror_child_hint != -1)
+			c = zio->io_mirror_child_hint % c + 1;
```

Comment:
This also looks wrong.  The hint should impact `mm_preferred`.  Looks like that's an array on ZoL, so we'll have to figure out what that means.  Maybe if the hint is set then vdev_mirror_child_select() can take a simpler code path (e.g. just set `mm_preferred[0] = ..._hint` and `mm_preferred_cnt = 1`).

Reply:
Yes, ```mm_preferred``` got changed to pointer (into ```mm_child[]```)  when the enhancements from FreeBSD were pulled in.  The highlighted lines above do, indirectly, cause ```mm_preferred``` at https://github.com/dweeezil/zfs/blob/illumos-7614/module/zfs/vdev_mirror.c#L156.

As I suggested in a comment above, it seems that as part of the FreeBSD work, ```mm_preferred``` is actually supposed to point to the slot in the ```mm_child[]``` array _one past``` the actual preferred entry.  I gleaned this from the logic in ```vdev_mirror_dva_select()``` in which the child number is pre-decremented.

All that said, I've not studied the whole FreeBSD mirror patch too closely but this certainly seems to be the way it's supposed to work.

Reply:
The `mm_preferred` array work from FreeBSD is quite confusing.  It actually points to an array of ints, of the same size as mm_child[] (i.e. there are `mm_children` entries in both mm_child[] and mm_preferred[]).  The code you referenced in vdev_mirror_map_alloc() is a very confusing way to allocate space for both mm_child[] and mm_preferred[].  mm_preferred is a pointer to the space after the last mm_child[].  vdev_mirror_map_size() ensures that enough space was allocated for this to work.

IMO, this is way more complicated than is necessary.  I would much prefer the simple solution of `mm_preferred = kmem_zalloc(mm_children * sizeof (int))`.

I looked at this some more last night and came to the conclusion that mm_preferred[] is essentially an ordered list of which children are preferred.  There may be multiple "preferred" children, up to the total number of children, which is why this is an array of size mm_children. 
 mm_preferred_cnt specifies how many preferred children there are (i.e. how many entries in mm_preferred should be considered).  But typically there's just one preferred child, in which case only mm_preferred[0] is used, and mm_preferred_cnt==1.

Reply:
Ugh, OK, I see now.  That is rather confusing and took a second look on my part, too.  It would be nice if there were at least some comments in the declaration of ```struct mirror_map``` to that it really has a pair of adjacent arrays, one of ```struct mirror_child``` followed by an array of ```int``` to which ```mm_preferred``` points.